### PR TITLE
presto-jdbc-driver: Resolve to CName pointing at DC hostname (A record) for SPN (Issues# 4828)

### LINKED ISSUE #4828: presto-jdbc-driver: Resolve to CName pointing at DC hostname (A record) for SPN
SPNEGO is used to setup authentication between client and server in presto-jdbc-driver using a service principal name(SPN).

if useCanonicalHostname is set to "false" then the hostname itself is used as servicehostname in SPN.
if useCanonicalHostname is set to "true" then the corresponding IP address is used as servicehostname in SPN.

Chrome established a convention that client resolves DNS CName entries first for all aliases until the resolver returns an A record.
So if an alias (CName) is pointing at DC hostnames (A records) then SPNEGO is attempted by Chrome using that alias as service principal name(SPN).
IP address are dynamic and cannot be considered SPNs irrespective of whether you are using AD or MIT kerberos.

If we can mimic this logic to use the alias (CName) pointing at DC hostnames( A records) then SPNEGO can be attempted using alias as SPN. 
This will enable usage of different aliases pointing to the same SPN without registering them separately as SPNs and this also helps with handling dynamic service IPs.

Can you please take a look? I would be happy to submit a PR with the code changes. Thank you!

### PR DESCRIPTION
**Pull request by ankuragarawal**
fixes #4828

### CODE REVIEW

File: presto-client/pom.xml [hunk 1]
```diff
@@ -75,10 +75,19 @@
             <scope>test</scope>
         </dependency>
 
+        <!--  TODO: Kindly suggest if another library is preferred. -->
+        <dependency>
```
Comment (Praveen2112) [inline presto-client/pom.xml:79]: Can we use TestNG ?

File: presto-client/src/main/java/io/prestosql/client/KerberosUtil.java [hunk 2]
```diff
@@ -13,13 +13,25 @@
...
 import java.util.Optional;
+import java.util.Properties;
 
 import static com.google.common.base.Strings.emptyToNull;
 import static com.google.common.base.Strings.nullToEmpty;
+import static java.util.Objects.requireNonNull;
 
 public final class KerberosUtil
 {
+    public static final String CNAME = "CNAME";
```
Comment (Praveen2112) [inline presto-client/src/main/java/io/prestosql/client/KerberosUtil.java:33]: We can have it as a private

File: presto-client/pom.xml (hunk 1 above)
Comment (ankuragarawal) [reply presto-client/pom.xml:79]: Hi Praveen,
I am using TestNG along with Mockito for mocking the dependencies. I did find a reference to mockito library in presto-bigquery as well so thought of adding it though if another library is preferred for mocking then I would be happy to replace it.  Let me know if you are good with this. Thank you!

File: presto-client/src/main/java/io/prestosql/client/KerberosUtil.java (hunk 2 above)
Comment (ankuragarawal) [reply presto-client/src/main/java/io/prestosql/client/KerberosUtil.java]: done!

**Comment by Praveen2112** [comment]
@cla-bot check

File: presto-client/src/main/java/io/prestosql/client/KerberosUtil.java [hunk 3]
```diff
@@ -32,4 +44,78 @@ private KerberosUtil() {}
...
+    private static String resolveInternal(String hostName, DirContext ictx)
+    {
+        try {
+            Attributes attributes = ictx.getAttributes(hostName, new String[]{CNAME, A_RECORD});
+            NamingEnumeration<? extends Attribute> namingEnumeration = attributes.getAll();
+
+            if (namingEnumeration.hasMore()) {
+                Attribute attribute = namingEnumeration.next();
+                String attributeId = attribute.getID();
+                switch (attributeId) {
```
Comment (Praveen2112) [inline presto-client/src/main/java/io/prestosql/client/KerberosUtil.java:90]: Can we inline `attributeId` ?

File: presto-client/src/main/java/io/prestosql/client/KerberosUtil.java [hunk 4]
```diff
@@ -32,4 +44,78 @@ private KerberosUtil() {}
...
+        catch (NamingException e) {
+            throw new ClientException("Failed to resolve host: " + hostName, e);
+        }
+        throw new ClientException("Failed to resolve host: " + hostName);
+    }
+
+    private static DirContext createDefaultDirContext()
+    {
+        Properties environment = new Properties();
+        environment.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.dns.DnsContextFactory");
```
Comment (Praveen2112) [inline presto-client/src/main/java/io/prestosql/client/KerberosUtil.java:109]:
```suggestion
        environment.setProperty(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.dns.DnsContextFactory");
```

File: presto-client/src/main/java/io/prestosql/client/KerberosUtil.java [hunk 5]
```diff
@@ -32,4 +44,78 @@ private KerberosUtil() {}
...
+        requireNonNull(host, "hostname is null");
+        try {
+            return resolveInternal(host, ictx);
+        }
+        catch (ClientException ignored) {
+        }
+        return host;
+    }
+
+    private static String resolveInternal(String hostName, DirContext ictx)
```
Comment (Praveen2112) [inline presto-client/src/main/java/io/prestosql/client/KerberosUtil.java:81]: Can we rename `ictx` to `context`

File: presto-client/src/main/java/io/prestosql/client/KerberosUtil.java (hunk 4 above)
Comment (Praveen2112) [reply presto-client/src/main/java/io/prestosql/client/KerberosUtil.java:109]: static import for `INITIAL_CONTEXT_FACTORY`

**Review by Praveen2112 — COMMENTED** [review]
Minor changes

**Comment by ankuragarawal** [comment]
@electrum Hi David, Can you please take another look at this feature/PR and let me know. I can resubmit it if it helps. Thanks!

**Comment by colebow** [comment]
:wave: @ankuragarawal - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

**Pull request closed without merging**